### DISCUSSION TITLE
Added: LDM Notes for December 15th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5549] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-12-15.md

## Agenda

1. Required parsing
2. Warnings for parameterless struct constructor

This is our last LDM for 2021: we'll be back again in January of 2022.

**Comment by DavidArno** [comment]
Probably a daft question here, but why can't the presence of field initializers simply trigger the automatic creation of a parameterless constructor? The whole issue of needing warnings just goes away then and that behaviour likely fits with what most developers are expecting the behaviour to be.

**Reply by YairHalberstadt** [reply]
Sounds sensible. This would be a breaking change, buy would match what I would expect.

**Reply by 333fred** [reply]
> why can't the presence of field initializers simply trigger the automatic creation of a parameterless constructor?

What happens for a record struct, where those initializers can depend on the primary constructor parameters?

**Reply by DavidArno** [reply]
For a record struct, either the constructor bypasses the rule around needing a this constructor, or it emits a this constructor with default values (both of which would have the same effect I think):
```c#
record struct R(int A)
{
    private int b = A;
    
    private int c = 1;
    
    public (int, int) GetBAndC() => (b, c);
    
    public R() : this(default) {} // auto-generated to ensure c = 1 is always called unless default(R) is used
}
```

**Reply by 333fred** [reply]
And what if the record struct instead looks like this?

```cs
record struct(MyClass c) 
{
    public string Str = c.ToString();
} 
```

**Reply by DavidArno** [reply]
And we reach that "oh, I hadn't though of that" moment that I've been waiting for since posting "_Probably a daft question here..._" :)

Personally I'd go for the breaking change here and have it throw a `NullReferenceException` when `new()` is called. I think the pain it would cause (which hopefully would be minimal given it's a change to a feature introduced with v10) would be worth it to create consistent behaviour with struct field initializers.

But "I'd go for the breaking change here" is dead easy for me to say as it's not my door that the angry mob would beat a path to should it prove a truly painful breaking change for them...

**Comment by Joe4evr** [comment]
For completeness: `required` parsing also touches on the discussion in #4460 (but does not link to it).
